storage/database/db_manager.go

```diff
@@ -371,10 +387,85 @@ func (dbm *databaseManager) IsParallelDBWrite() bool {
 	return dbm.config.ParallelDBWrite
 }
 
+func (dbm *databaseManager) InMigration() bool {
+	return dbm.inMigration
+}
+
 func (dbm *databaseManager) NewBatch(dbEntryType DBEntryType) Batch {
+	if dbEntryType == StateTrieDB && dbm.inMigration {
+		return dbm.newStateTrieDB.NewBatch()
+	}
 	return dbm.getDatabase(dbEntryType).NewBatch()
 }
 
+func (dbm *databaseManager) getStateTrieMigrationInfo() (bool, string) {
+	miscDB := dbm.getDatabase(MiscDB)
+
+	enc, _ := miscDB.Get(migrationStatusKey)
+	if len(enc) == 0 {
+		return false, ""
+	}
+
+	var status uint64
+	if err := rlp.DecodeBytes(enc, &status); err != nil {
+		logger.Error("Failed to decode migration status", "err", err)
+		return false, ""
+	}
+
+	if status == 0 {
+		return false, ""
+	}
+
+	enc, _ = miscDB.Get(databaseDirKey(uint64(StateTrieDB)))
+	return true, string(enc)
+}
+
+func createStateTrieMigrationDB(dbc *DBConfig, newDBDir string) Database {
+	oldDBConfig := dbc
+	newDBConfig := getDBEntryConfig(oldDBConfig, StateTrieDB)
+	newDBConfig.Dir = newDBDir
+	var newDB Database
+	var err error
+	if newDBConfig.NumStateTriePartitions > 1 {
+		newDB, err = newPartitionedDB(newDBConfig, StateTrieDB, newDBConfig.NumStateTriePartitions)
+	} else {
+		newDB, err = newDatabase(newDBConfig, StateTrieDB)
+	}
+	if err != nil {
+		logger.Crit("Failed to create a new database for state trie migration", "err", err)
+	}
+	logger.Info("Created a new database for state trie migration", "newStateTrieDB", newDBConfig.Dir)
+	return newDB
+}
+
+func (dbm *databaseManager) SetStateTrieMigrationDB(blockNum uint64) {
```

Comment:
Is it okay even if the process is terminated during the execution of this function?

Reply:
"okay" can be understood two different ways. If terminated during `SetStateTrieMigrationDB`,
1. It does not make problem, but doesn't start migration process automatically after re-run neither
2. It does not make problem, and start migration process after re-run

We'll take the first approach with the assumption there is only little possibility that the node is terminated during the execution of `SetStateTrieMigrationDB`. **Therefore, if it is failed during the execution of the function, migration process should be re-triggered by outer process, not by itself.**